### PR TITLE
[DONT MERGE] Added Near Cache preloader to IMap on Hazelcast members

### PR DESCRIPTION
**Pull request by Donnerbart**
* implemented Near Cache preloader in `NearCachedMapProxyImpl`
* added XML configuration to Hazelcast and Spring XSD
* added configuration parsing to `XMLConfigBuilder` and `BeanDefinitionParser`
* removed client check from `ConfigValidator`
* moved test resource files to core module

I've added the Near Cache preloader to IMap on Hazelcast members, to make the feature more complete. This will also increase the Sonar code coverage of the `NearCachePreloader` (cross-module issue, just tested via clients now).

I also added the preloader configuration to Spring, which was a bit difficult for the client part. The `NearCacheConfig` has `getPreloaderConfig()` and `setPreloaderConfig()`, but the XML element is named `preloader`, which breaks automatic parsing of the `BeanDefinition`. So manual parsing code was added.

Fixes https://github.com/hazelcast/hazelcast/issues/9771

### CODE REVIEW

**Comment by Donnerbart** [comment]
run-lab-run

File: hazelcast-spring/src/main/java/com/hazelcast/spring/HazelcastClientBeanDefinitionParser.java [hunk 1]
```diff
@@ -243,7 +244,26 @@ private void handleLoadBalancer(Node node) {
         }
 
         private void handleNearCache(Node node) {
-            createAndFillListedBean(node, NearCacheConfig.class, "name", nearCacheConfigMap, "name");
+            BeanDefinitionBuilder nearCacheConfigBuilder = createBeanBuilder(NearCacheConfig.class);
```
Comment (Donnerbart) [inline hazelcast-spring/src/main/java/com/hazelcast/spring/HazelcastClientBeanDefinitionParser.java]: The `createAndFillListedBean()` couldn't handle the `preloader` vs. `preloaderConfig` mismatch, so I added a custom parsing code here.

File: hazelcast-spring/src/test/resources/com/hazelcast/spring/node-client-applicationContext-hazelcast.xml [hunk 2]
```diff
@@ -22,7 +22,8 @@
        xsi:schemaLocation="http://www.springframework.org/schema/beans
 		http://www.springframework.org/schema/beans/spring-beans-2.5.xsd
 		http://www.hazelcast.com/schema/spring
-		http://www.hazelcast.com/schema/spring/hazelcast-spring.xsd">
+		hazelcast-spring-3.9.xsd
```
Comment (Donnerbart) [inline hazelcast-spring/src/test/resources/com/hazelcast/spring/node-client-applicationContext-hazelcast.xml]: We specify the XSD version in the member XML as well and for me this fixes the parsing (somehow `hazelcast-spring.xsd` doesn't get resolved to 3.9 correctly).

File: hazelcast/src/main/java/com/hazelcast/map/impl/proxy/NearCachedMapProxyImpl.java [hunk 3]
```diff
@@ -92,7 +93,8 @@ public void initialize() {
         super.initialize();
 
         mapNearCacheManager = mapServiceContext.getMapNearCacheManager();
-        nearCache = mapNearCacheManager.getOrCreateNearCache(name, mapConfig.getNearCacheConfig());
+        IMapDataStructureAdapter<K, V> adapter = new IMapDataStructureAdapter<K, V>(this);
+        nearCache = mapNearCacheManager.getOrCreateNearCache(name, mapConfig.getNearCacheConfig(), adapter);
```
Comment (Donnerbart) [inline hazelcast/src/main/java/com/hazelcast/map/impl/proxy/NearCachedMapProxyImpl.java]: This is all the code it needs to implement the Near Cache preloader for `IMap` on a Hazelcast member :)

File: hazelcast/src/test/java/com/hazelcast/internal/config/ConfigValidatorTest.java [hunk 4]
```diff
@@ -269,7 +269,7 @@ public void test_checkNearCacheConfig_withPreLoaderConfig_onClients() {
         checkNearCacheConfig(MAP_NAME, nearCacheConfig, true);
     }
 
-    @Test(expected = IllegalArgumentException.class)
+    @Test
```
Comment (Donnerbart) [inline hazelcast/src/test/java/com/hazelcast/internal/config/ConfigValidatorTest.java]: This was the test which failed, when the `NearCachePreloaderConfig` was configured for a member -> needs to pass now.

File: hazelcast/src/test/java/com/hazelcast/map/impl/nearcache/MapNearCachePreLoaderTest.java [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+import static com.hazelcast.map.impl.nearcache.MapInvalidationListener.createInvalidationEventHandler;
+import static java.util.Arrays.asList;
+
+/**
+ * Basic Near Cache tests for {@link IMap} on Hazelcast members.
+ */
+@RunWith(Parameterized.class)
+@UseParametersRunnerFactory(HazelcastParametersRunnerFactory.class)
+@Category({ParallelTest.class, QuickTest.class})
+public class MapNearCachePreLoaderTest extends AbstractNearCachePreloaderTest<Data, String> {
```
Comment (Donnerbart) [inline hazelcast/src/test/java/com/hazelcast/map/impl/nearcache/MapNearCachePreLoaderTest.java]: This is the test for the Near Cache preloader on `IMap` for Hazelcast members.

**Comment by ahmetmircik** [comment]
will we have tests for NATIVE maps?

**Comment by Donnerbart** [comment]
Yes, should be very easy to provide the HD test (override `getConfig()` and provide a new parameter set).

**Comment by tombujok** [comment]
@ahmetmircik @Donnerbart let us discuss this on Monday. It's feature freeze phase, this is effectively adding a preloader to members, which is a new feature...

**Comment by Donnerbart** [comment]
The feature is itself is not new, neither is its code. It's just enabled in a missing part to solve the confusion in which data structures the Near Cache preloader is available.

**Comment by tombujok** [comment]
@Donnerbart To me this sentences is the key `I've added the Near Cache preloader to IMap on Hazelcast members, to make the feature more complete. `. Earlier there was no preloader on the members. I knot the code existed, but was not used on members. Did you test it in simulator? What are the consequences performance and memory wise? Are you sure it will not cause OOME, what about bouncing nodes or stalling partition threads? That is the reason we have feature freeze, not to introduce new mechanisms to the execution path... Adding a new feature in the feature freeze period will require an extra permission, I'm not super fond of that...

**Comment by Donnerbart** [comment]
The preloader code is exactly the same as on the client and it is just executed if you explicitly configure it. So by just adding the feature to members, there will nothing happen, beside that the feature is accessible to more users. The consequences of using it are the same as on clients regarding memory and performance. The preloader tests are unified, so we'll have exactly the same coverage like we have on the clients.

The direct consequences are a `IMapDataStructureAdapter` wrapper class per map proxy. This could be optimized, if it's an issue (would do the same on the client then). Beside that there were only two classes changed to activate the feature, all other classes are config or test related.

**Comment by Donnerbart** [comment]
Vassilis had a very good comment on this: there might be a conflict between MapLoader and the Near Cache preloader on members. On client we made clear that it's the responsibility that the cluster has the data loaded, which should be preloaded for the Near Cache. On the members the user cannot control this, so we have to check this. This is a blocker for 3.9, so I'm moving this to 3.10.

I pulled out the Spring client config to this PR: https://github.com/hazelcast/hazelcast/pull/10980

**Comment by jerrinot** [comment]
@Donnerbart: what plans you have with this PR?

**Pull request closed without merging**